src/lib.rs

```diff
@@ -39,12 +39,12 @@ fn _decode_raw<'a>(
     preserve_cache_info: bool,
     rlp_val: rlp::Rlp,
     py: pyo3::Python<'a>,
-) -> Result<(ListOrBytes<'a>, Option<ListOrBytes<'a>>), PyErr> {
+) -> PyResult<(ListOrBytes<'a>, Option<ListOrBytes<'a>>)> {
```

Comment:
Could we put a name on these values?

```rust
struct SomeName{
  decoded_val: ListOrBytes,
  rlp_val: Option<ListOrBytes>
}
```

I'm having trouble figuring out what's being returned here

Reply:
Yeah, I guess we could. Let me illustrate it with this test:

https://github.com/cburgdorf/rusty-rlp/blob/41af02048f60a8ce4f4193a03c4b8f3a04b4491f/python_tests.py#L143-L145

The first line are the RLP bytes that we want to decode, so it corresponds to the `rlp_val` that goes into `_decode_raw`. The second line from this test data is the decoded data so it corresponds to `decoded_val` in your struct. And the last line represents all the *intermediate* RLP values (and what they resolve to) that were "hidden" inside the RLP strings. You can think of it as additional information that we *optionally* preserve for the caller (if `preserve_cache_info` is `true`) if they want to *recursively* cache RLP values. 

To give a more practical example. Imagine you are decoding a byte string `b'deadbeefc0ffe'` that happens to resolve to `[<Block1>, <Block2>, <Block3>]`. Now if you only have the input and output value then you can cache that `b'deadbeefc0ffe'` resolves to  `[<Block1>, <Block2>, <Block3>]` but you can not cache the RLP string that resolves to `<Block1>`, `<Block2>` or `<Block3>` individually. This is where `preserve_cache_info` comes in because it gives you detailed information about all the **inner** RLPs and what they resolve to.

This is a [requirement that comes from the pyrlp libary](https://github.com/ethereum/pyrlp/blob/5a49f8e5bb6835f00b8c314641107c9deeeec9f2/rlp/codec.py#L230) that uses `rusty-rlp`. So, the format that this takes was already used by pyrlp and I just tried to match it.

So, I guess we could call the struct something like:

```rust
struct DecodingInfo{
  decoded_val: ListOrBytes,
  cache_info: Option<ListOrBytes>
}
```

Notice that doing this extra book keeping has its price and hence the code tries to circumvent it entirely if `preserve_cache_info` is `false`. It's a pretty niche feature tbh but the p2p library of Trinity uses it and so I needed to support it.

Reply:
thanks for the info, makes sense :+1: